File: daemon/src/assets/catalog.ts
import * as fs from 'fs';
import * as path from 'path';

import {
  flattenManifest,
  type FurnitureAsset,
  type FurnitureManifest,
} from '../../../shared/assets/manifestUtils.js';

export type { FurnitureAsset };

export interface BuiltCatalog {
  assets: FurnitureAsset[];
  /** groupId → ordered list of asset IDs forming a rotation cycle */
  rotationGroups: Map<string, string[]>;
  /** assetId → asset ID of the toggled state sibling */
  stateGroups: Map<string, string>;
}

/**
 * Scan a furniture root directory (e.g. `webview-ui/public/assets/furniture`).
 * Each immediate subdirectory is expected to contain a `manifest.json`.
 * Subdirectories missing a manifest are silently skipped.
 */
export function scanFurnitureDir(furnitureDir: string): FurnitureAsset[] {
  if (!fs.existsSync(furnitureDir)) return [];

  const entries = fs.readdirSync(furnitureDir, { withFileTypes: true });
  const assets: FurnitureAsset[] = [];

  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const manifestPath = path.join(furnitureDir, entry.name, 'manifest.json');
    if (!fs.existsSync(manifestPath)) continue;

    let manifest: FurnitureManifest;
    try {
      manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf8')) as FurnitureManifest;
    } catch {
      continue; // malformed manifest — skip
    }

    const inherited = {
      groupId: manifest.id,
      name: manifest.name,
      category: manifest.category,
      canPlaceOnWalls: manifest.canPlaceOnWalls ?? false,
      canPlaceOnSurfaces: manifest.canPlaceOnSurfaces ?? false,
      backgroundTiles: manifest.backgroundTiles ?? 0,
    };

    if (manifest.type === 'asset') {
      // Single-asset furniture (no rotation / state groups)
      assets.push({
        id: manifest.id,
        name: manifest.name,
        label: manifest.name,
        category: manifest.category,
        file: manifest.file ?? `${manifest.id}.png`,
        width: manifest.width ?? 0,
        height: manifest.height ?? 0,
        footprintW: manifest.footprintW ?? 1,
        footprintH: manifest.footprintH ?? 1,
        isDesk: manifest.category === 'desks',
        canPlaceOnWalls: manifest.canPlaceOnWalls ?? false,
        canPlaceOnSurfaces: manifest.canPlaceOnSurfaces ?? false,
        backgroundTiles: manifest.backgroundTiles ?? 0,
      });
    } else if (manifest.type === 'group' && Array.isArray(manifest.members)) {
      for (const member of manifest.members) {
        assets.push(...flattenManifest(member, inherited));
      }
    }
  }

  return assets;
}

/**
 * Build rotation and state group maps from a flat asset list.
 *
 * Rotation groups: assets sharing a `groupId` with different `orientation`
 * values. Ordering follows the ORIENTATION_ORDER list so rotation cycles
 * are predictable regardless of manifest order.
 *
 * State groups: within a rotation group (same `groupId` + `orientation`),
 * assets with `state: 'on'` and `state: 'off'` form a toggle pair.
 * The map is bidirectional: on→off and off→on.
 */
export function buildGroups(assets: FurnitureAsset[]): BuiltCatalog {
  // groupId → orientation → asset[]
  const byGroupOrient = new Map<string, Map<string, FurnitureAsset[]>>();

  for (const asset of assets) {
    if (!asset.groupId) continue;
    let byOrient = byGroupOrient.get(asset.groupId);
    if (!byOrient) {
      byOrient = new Map();
      byGroupOrient.set(asset.groupId, byOrient);
    }
    const orient = asset.orientation ?? '';
    const list = byOrient.get(orient) ?? [];
    list.push(asset);
    byOrient.set(orient, list);
  }

  const rotationGroups = new Map<string, string[]>();
  const stateGroups = new Map<string, string>();

  for (const [groupId, byOrient] of byGroupOrient) {
    // Rotation: gather one representative per orientation (prefer state='off' or no state)
    const orientKeys = [...byOrient.keys()];
    if (orientKeys.length > 1 || (orientKeys.length === 1 && orientKeys[0] !== '')) {
      const cycle = orientKeys.sort(compareOrientations).map((orient) => {
        const group = byOrient.get(orient)!;
        // Pick off/unset state as the rotation representative
        const rep = group.find((a) => !a.state || a.state === 'off') ?? group[0]!;
        return rep.id;
      });
      if (cycle.length > 1) {
        rotationGroups.set(groupId, cycle);
      }
    }

    // State: within each orientation bucket, wire on↔off pairs
    for (const group of byOrient.values()) {
      const off = group.find((a) => !a.state || a.state === 'off');
      const on = group.find((a) => a.state === 'on');
      if (off && on) {
        stateGroups.set(off.id, on.id);
        stateGroups.set(on.id, off.id);
      }
    }
  }

  return { assets, rotationGroups, stateGroups };
}

const ORIENTATION_ORDER = ['front', 'back', 'side', 'left', 'right', 'top'];

function compareOrientations(a: string, b: string): number {
  const ai = ORIENTATION_ORDER.indexOf(a);
  const bi = ORIENTATION_ORDER.indexOf(b);
  if (ai === -1 && bi === -1) return a.localeCompare(b);
  if (ai === -1) return 1;
  if (bi === -1) return -1;
  return ai - bi;
}

File: client/src/office/catalog.rs
#![allow(dead_code)]

use std::collections::BTreeMap;

use serde_json::Value;

// ── Catalog entry ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct FurnitureCatalogEntry {
    pub id: String,
    pub name: String,
    pub category: String,
    pub footprint_w: i32,
    pub footprint_h: i32,
    pub is_desk: bool,
    pub can_place_on_walls: bool,
    pub can_place_on_surfaces: bool,
    pub background_tiles: i32,
    pub orientation: Option<String>,
    pub group_id: Option<String>,
    /// Pixel height of the sprite — used for zY calculation (Day 4-7 approximation).
    pub sprite_height: i32,
    /// "on" | "off" | None
    pub state: Option<String>,
}

// ── Catalog ───────────────────────────────────────────────────────────────────

pub struct FurnitureCatalog {
    pub entries: BTreeMap<String, FurnitureCatalogEntry>,
    /// groupId → ordered rotation cycle (sorted by ORIENTATION_ORDER)
    pub rotation_groups: BTreeMap<String, Vec<String>>,
    /// assetId → sibling state assetId (bidirectional: on↔off)
    pub state_groups: BTreeMap<String, String>,
}

const ORIENTATION_ORDER: &[&str] = &["front", "back", "side", "left", "right", "top"];

fn orientation_rank(o: &str) -> usize {
    ORIENTATION_ORDER.iter().position(|&x| x == o).unwrap_or(ORIENTATION_ORDER.len())
}

impl FurnitureCatalog {
    pub fn empty() -> Self {
        Self {
            entries: BTreeMap::new(),
            rotation_groups: BTreeMap::new(),
            state_groups: BTreeMap::new(),
        }
    }

    /// Build catalog from a WorldSnapshot `Value` (from HelloAck).
    pub fn from_wire(world: &Value) -> Self {
        let Some(arr) = world
            .get("assets")
            .and_then(|a| a.get("catalog"))
            .and_then(|c| c.as_array())
        else {
            return Self::empty();
        };

        let mut entries: BTreeMap<String, FurnitureCatalogEntry> = BTreeMap::new();

        for item in arr {
            let id = match item.get("id").and_then(|v| v.as_str()) {
                Some(s) if !s.is_empty() => s.to_string(),
                _ => continue,
            };
            entries.insert(
                id.clone(),
                FurnitureCatalogEntry {
                    id,
                    name: item.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                    category: item.get("category").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                    footprint_w: item.get("footprintW").and_then(|v| v.as_i64()).unwrap_or(1) as i32,
                    footprint_h: item.get("footprintH").and_then(|v| v.as_i64()).unwrap_or(1) as i32,
                    is_desk: item.get("isDesk").and_then(|v| v.as_bool()).unwrap_or(false),
                    can_place_on_walls: item.get("canPlaceOnWalls").and_then(|v| v.as_bool()).unwrap_or(false),
                    can_place_on_surfaces: item.get("canPlaceOnSurfaces").and_then(|v| v.as_bool()).unwrap_or(false),
                    background_tiles: item.get("backgroundTiles").and_then(|v| v.as_i64()).unwrap_or(0) as i32,
                    orientation: item.get("orientation").and_then(|v| v.as_str()).map(|s| s.to_string()),
                    group_id: item.get("groupId").and_then(|v| v.as_str()).map(|s| s.to_string()),
                    sprite_height: item.get("height").and_then(|v| v.as_i64()).unwrap_or(16) as i32,
                    state: item.get("state").and_then(|v| v.as_str()).map(|s| s.to_string()),
                },
            );
        }

        let (rotation_groups, state_groups) = build_groups(&entries);
        Self { entries, rotation_groups, state_groups }
    }

    pub fn get(&self, id: &str) -> Option<&FurnitureCatalogEntry> {
        self.entries.get(id)
    }

    pub fn get_on_state(&self, id: &str) -> Option<&str> {
        self.state_groups.get(id).map(|s| s.as_str())
    }
}

fn build_groups(
    entries: &BTreeMap<String, FurnitureCatalogEntry>,
) -> (BTreeMap<String, Vec<String>>, BTreeMap<String, String>) {
    // groupId → orientation → Vec<asset_id>
    let mut by_group_orient: BTreeMap<String, BTreeMap<String, Vec<String>>> = BTreeMap::new();

    for (id, entry) in entries {
        if let Some(ref gid) = entry.group_id {
            by_group_orient
                .entry(gid.clone())
                .or_default()
                .entry(entry.orientation.clone().unwrap_or_default())
                .or_default()
                .push(id.clone());
        }
    }

    let mut rotation_groups: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let mut state_groups: BTreeMap<String, String> = BTreeMap::new();

    for (gid, by_orient) in &by_group_orient {
        let mut orient_keys: Vec<&String> = by_orient.keys().collect();

        // Rotation groups (multiple orientations, or single non-empty orientation)
        if orient_keys.len() > 1 || (orient_keys.len() == 1 && !orient_keys[0].is_empty()) {
            orient_keys.sort_by(|a, b| {
                orientation_rank(a).cmp(&orientation_rank(b)).then(a.cmp(b))
            });

            let cycle: Vec<String> = orient_keys
                .iter()
                .filter_map(|orient| {
                    let group = by_orient.get(*orient)?;
                    // Prefer non-"on" state as rotation representative
                    let rep = group
                        .iter()
                        .find(|id| entries.get(*id).and_then(|e| e.state.as_deref()) != Some("on"))
                        .or_else(|| group.first())?;
                    Some(rep.clone())
                })
                .collect();

            if cycle.len() > 1 {
                rotation_groups.insert(gid.clone(), cycle);
            }
        }

        // State groups: wire on↔off within each orientation bucket
        for group in by_orient.values() {
            let off = group.iter().find(|id| {
                entries.get(*id).and_then(|e| e.state.as_deref()) != Some("on")
            });
            let on = group.iter().find(|id| {
                entries.get(*id).and_then(|e| e.state.as_deref()) == Some("on")
            });
            if let (Some(off_id), Some(on_id)) = (off, on) {
                state_groups.insert(off_id.clone(), on_id.clone());
                state_groups.insert(on_id.clone(), off_id.clone());
            }
        }
    }

    (rotation_groups, state_groups)
}
